CSharp/Library/Microsoft.Bot.Builder/Dialogs/PromptDialog.cs

```diff
@@ -577,39 +635,59 @@ public PromptChoice(IEnumerable<T> options, string prompt, string retry, int att
+            /// <param name="choices">Dictionary of the options to choose from with the synonyms.</param>
+            public PromptChoice(PromptOptions<T> promptOptions, IDictionary<T, IEnumerable<T>> choices = null)
                 : base(promptOptions)
             {
                 SetField.CheckNull(nameof(promptOptions.Options), promptOptions.Options);
-            }
-
-            public virtual Tuple<bool, int> ScoreMatch(T option, string input)
-            {
-                var trimmed = input.Trim();
-                var text = option.ToString();
-                bool occurs = text.IndexOf(trimmed, StringComparison.CurrentCultureIgnoreCase) >= 0;
-                bool equals = text == trimmed;
-                return occurs
-                    ? Tuple.Create(equals, trimmed.Length)
-                    : null;
+                if (choices == null)
+                {
+                    this.choices = new Dictionary<T, IEnumerable<T>>();
+                    promptOptions.Options.ToList().ForEach(x => this.choices.Add(x, null));
+                }
+                else
+                {
+                    this.choices = choices;
+                }
             }
 
             protected override bool TryParse(IMessageActivity message, out T result)
             {
                 if (!string.IsNullOrWhiteSpace(message.Text))
                 {
-                    var scores = from option in this.promptOptions.Options
-                                 let score = ScoreMatch(option, message.Text)
-                                 select new { score, option };
+                    var entityMatches = PromptRecognizers.RecognizeChoices(message.Text, choices);
+                    var cardinalMatches = PromptRecognizers.RecognizeNumbers(message, new PromptRecognizeNumbersOptions { IntergerOnly = true, MinValue = 0, MaxValue = choices.Count - 1 });
+                    var ordinalMatches = PromptRecognizers.RecognizeOrdinals(message);
+
+                    var entityWinner = entityMatches.MaxBy(x => x.Score) ?? new RecognizeEntity<T>();
+                    var cardinalWinner = cardinalMatches.MaxBy(x => x.Score) ?? new RecognizeEntity<double>();
+                    var ordinalWinner = ordinalMatches.MaxBy(x => x.Score) ?? new RecognizeEntity<long>();
 
-                    var winner = scores.MaxBy(s => s.score);
-                    if (winner.score != null)
+                    if (entityWinner.Entity != null && entityWinner.Score >= Math.Max(cardinalWinner.Score, ordinalWinner.Score))
+                    {
+                        result = entityWinner.Entity;
+                        return true;
+                    }
+                    else if (Math.Max(cardinalWinner.Score, ordinalWinner.Score) > 0)
                     {
-                        result = winner.option;
+                        var index = (int)(cardinalWinner.Score > ordinalWinner.Score ? cardinalWinner.Entity : ordinalWinner.Entity);
```

Comment:
Replaced current logic with the latest one in Node. Also, fixed an issue related to reverse ordinals and created a PR too for Node #2623